### PR TITLE
add code for catalog and list tag API

### PR DESCRIPTION
**Pull request by wy65701436**
Signed-off-by: wang yan <wangyan@vmware.com>

### CODE REVIEW

File: src/server/registry/catalog/catalog.go [hunk 1]
```diff
@@ -14,19 +14,109 @@
...
+
+func (h *handler) get(w http.ResponseWriter, req *http.Request) {
+	var withPagination = true
+
+	reqQ := req.URL.Query()
+	lastEntry := reqQ.Get("last")
+	maxEntries, err := strconv.Atoi(reqQ.Get("n"))
+	if err != nil || maxEntries < 0 {
+		err := ierror.New(err).WithCode(ierror.BadRequestCode).WithMessage("the N must be a positive int type")
+		reg_error.Handle(w, req, err)
```
Comment (ywk253100) [inline src/server/registry/catalog/catalog.go]: The response data should be in the format of `catalogAPIResponse`

File: src/server/registry/catalog/catalog.go [hunk 2]
```diff
@@ -14,19 +14,109 @@
...
+
+	var repos []string
+	// get all repositories
+	total, repoRecords, err := h.repoMgr.List(req.Context(), nil)
+	if err != nil {
+		error.Handle(w, req, err)
+		return
+	}
+	if total <= 0 {
+		h.sendResponse(w, req, repos)
```
Comment (ywk253100) [inline src/server/registry/catalog/catalog.go]: The response data should be in the format of `catalogAPIResponse`

File: src/server/registry/catalog/catalog.go [hunk 3]
```diff
@@ -14,19 +14,109 @@
...
+	}
+	for _, r := range repoRecords {
+		repos = append(repos, r.Name)
+	}
+	sort.Strings(repos)
+
+	// with "last", if last page, no link header
+	// without "last", if one page covers "n", no link header
+	if lastEntry != "" {
+		lastEntryIndex := util.IndexString(repos, lastEntry)
```
Comment (ywk253100) [inline src/server/registry/catalog/catalog.go]: Check `if lastEntryIndex == -1` ?

File: src/server/registry/tag/tag.go [hunk 4]
```diff
@@ -15,22 +15,165 @@
...
+	repoID, err = h.getRepoID(req)
+	if err != nil {
+		reg_error.Handle(w, req, err)
+		return
+	}
+
+	// get tags ...
+	total, tags, err := h.tagMgr.List(req.Context(), &q.Query{
+		Keywords: map[string]interface{}{
+			"repository_id": repoID,
```
Comment (ywk253100) [inline src/server/registry/tag/tag.go]: Use `RepositoryID` instead. `repository_id` is column name of the database table, it shouldn't be aware by manager level

File: src/server/registry/tag/tag.go [hunk 5]
```diff
@@ -15,22 +15,165 @@
...
+		},
+	})
+	if err != nil {
+		return 0, err
+	}
+	if total <= 0 {
+		err := ierror.New(errors.New("repositoryNotFound")).WithCode(ierror.NotFoundCode)
+		return 0, err
+	}
+	if total >= 1 {
```
Comment (ywk253100) [inline src/server/registry/tag/tag.go]: The `if` is useless

File: src/server/registry/tag/tag.go [hunk 6]
```diff
@@ -15,22 +15,165 @@
...
+	total, repoRecord, err := h.repoMgr.List(req.Context(), &q.Query{
+		Keywords: map[string]interface{}{
+			"name": h.repositoryName,
+		},
+	})
+	if err != nil {
+		return 0, err
+	}
+	if total <= 0 {
+		err := ierror.New(errors.New("repositoryNotFound")).WithCode(ierror.NotFoundCode)
```
Comment (ywk253100) [inline src/server/registry/tag/tag.go]: This is where the error generated, so I think the original cause should be nil:
`ierror.New(nil).WithCode(xxx).WithMessage(xxx)`?
Specify what repository name is in the massage

File: src/server/registry/tag/tag.go (hunk 5 above)
Comment (wy65701436) [reply src/server/registry/tag/tag.go]: udpated.

File: src/server/registry/tag/tag.go (hunk 6 above)
Comment (wy65701436) [reply src/server/registry/tag/tag.go]: updated.

File: src/server/registry/catalog/catalog.go (hunk 2 above)
Comment (wy65701436) [reply src/server/registry/catalog/catalog.go]: yes, it is.

File: src/server/registry/tag/tag.go (hunk 4 above)
Comment (wy65701436) [reply src/server/registry/tag/tag.go]: did

File: src/server/registry/catalog/catalog.go [hunk 7]
```diff
@@ -14,19 +14,117 @@
...
+	}
+}
+
+func (h *handler) get(w http.ResponseWriter, req *http.Request) {
+	var maxEntries int
+	var err error
+
+	reqQ := req.URL.Query()
+	lastEntry := reqQ.Get("last")
+	withN := reqQ.Get("n") == ""
```
Comment (heww) [inline src/server/registry/catalog/catalog.go]: `withN := reqQ.Get("n") != ""` ?

Comment (wy65701436) [reply src/server/registry/catalog/catalog.go]: good catch.

File: src/server/registry/catalog/catalog.go [hunk 8]
```diff
@@ -14,19 +14,117 @@
...
+		return
+	}
+	for _, r := range repoRecords {
+		repoNames = append(repoNames, r.Name)
+	}
+	sort.Strings(repoNames)
+	if !withN {
+		h.sendResponse(w, req, repoNames)
+		return
+	}
```
Comment (heww) [inline src/server/registry/catalog/catalog.go]: This will cause the API returns all repositories when `last` in the query but `n` not in it.

Comment (wy65701436) [reply src/server/registry/catalog/catalog.go]: we will not handle the query only with `last`

**Pull request merged**